src/derivative.jl

```diff
@@ -0,0 +1,38 @@
+default_derivative_algorithm() = "exact"
+
+function derivative(
+  ψ::AbstractITensorNetwork, vertices::Vector; alg=default_derivative_algorithm(), kwargs...
+)
+  return derivative(Algorithm(alg), ψ, vertices; kwargs...)
+end
+
+function derivative(
+  ::Algorithm"exact",
+  ψ::AbstractITensorNetwork,
+  vertices::Vector;
+  contraction_sequence_alg="optimal",
+  kwargs...,
+)
+  ψ_reduced = Vector{ITensor}(subgraph(ψ, vertices))
+  sequence = contraction_sequence(ψ_reduced; alg=contraction_sequence_alg)
+  return ITensor[contract(ψ_reduced; sequence, kwargs...)]
+end
+
+function derivative(
+  ::Algorithm"bp",
+  ψ::AbstractITensorNetwork,
+  verts::Vector;
+  (bp_cache!)=nothing,
+  update_bp_cache=isnothing(bp_cache!),
+  bp_cache_update_kwargs=default_cache_update_kwargs(bp_cache!),
```

Comment:
Maybe just `cache!`, `update_cache`, and `update_kwargs`? Specifying it is `bp` seems redundant since it's already been specified that it is the `bp` through the algorithm selection.